src/librustc/mir/interpret/mod.rs

```diff
@@ -238,6 +242,127 @@ pub fn specialized_decode_alloc_id<
     }
 }
 
+// Used to avoid infinite recursion when decoding recursive allocations.
+pub type DecodingSessionId = NonZeroU32;
+
+// FIXME: This could be represented as an AtomicU64, letting us get rid of the
+//        Mutex in AllocIdDecoder
+#[derive(Copy, Clone)]
+enum State {
+    Empty,
+    InProgress(DecodingSessionId, AllocId),
+    Done(AllocId),
+    Poisoned,
+}
+
+pub struct AllocIdDecoder {
+    // For each AllocId we keep track of which decoding state it's currently in.
+    decoding_state: Mutex<Vec<State>>,
+    // The offsets of each allocation in the data stream.
+    data_offsets: Vec<u32>,
+}
+
+impl AllocIdDecoder {
+
+    // FIXME: Don't expose this.
+    pub fn new_decode_session() -> DecodingSessionId {
+        static DECODER_SESSION_ID: AtomicU32 = AtomicU32::new(0);
+        let counter = DECODER_SESSION_ID.fetch_add(1, Ordering::SeqCst);
+
+        // Make sure this is never zero
+        DecodingSessionId::new((counter & 0x7FFFFFFF) + 1).unwrap()
+    }
+
+    pub fn new(data_offsets: Vec<u32>) -> AllocIdDecoder {
+        let decoding_state: Vec<_> = ::std::iter::repeat(State::Empty)
+            .take(data_offsets.len())
+            .collect();
+
+        AllocIdDecoder {
+            decoding_state: Mutex::new(decoding_state),
+            data_offsets,
+        }
+    }
+
+    pub fn decode_alloc_id<'a, 'tcx, D>(&self,
+                                        decoder: &mut D,
+                                        session_id: DecodingSessionId)
+                                        -> Result<AllocId, D::Error>
+        where D: TyDecoder<'a, 'tcx>,
+              'tcx: 'a,
+    {
+        // Read the index of the allocation
+        let idx = decoder.read_u32()? as usize;
+
+        // Check the decoding state, see if it's already decoded or if we should
+        // decode it here.
+        let alloc_id = loop {
+            let mut decoding_state = self.decoding_state.lock();
+            let entry = &mut decoding_state[idx];
+
+            match *entry {
+                State::Done(alloc_id) => {
+                    return Ok(alloc_id);
+                }
+                State::Empty => {
+                    // We are allowed to decode
+                    let alloc_id = decoder.tcx().alloc_map.lock().reserve();
+                    *entry = State::InProgress(session_id, alloc_id);
+                    break alloc_id;
+                }
+                State::InProgress(other_session_id, alloc_id) => {
+                    if other_session_id == session_id {
+                        // Don't recurse.
+                        return Ok(alloc_id);
+                    } else {
+                        // Just spin. We could also decode concurrently.
```

Comment:
ideally we'd turn this into a work stealing algo for allocations with loads of relocations... someday

Reply:
Yes, with Rayon it might not even be that hard.